### PR TITLE
Regulator Functionality for Exporter

### PR DESCRIPTION
**Pull request by InusualZ**
It does not handle the GUI correctly

### CODE REVIEW

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java [hunk 1]
```diff
@@ -59,7 +78,21 @@ public void updateNode() {
...
 
                         if (!slot.isEmpty()) {
-                            ItemStack took = network.extractItem(slot, upgrades.getItemInteractCount(), compare, true);
+                            int stackSize = upgrades.getItemInteractCount();
+
+                            if (regulator) {
+                                for(int index = 0; i< handler.getSlots(); i++) {
+                                    if (slot.isItemEqual(handler.getStackInSlot(index))) {
+                                        if (handler.getStackInSlot(index).getCount() >= slot.getCount()) {
+                                            return;
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: This seems like a very odd place to return. Shouldn't it be some kind of break

File: src/main/java/com/raoulvdberge/refinedstorage/gui/sidebutton/SideButtonRegulator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.raoulvdberge.refinedstorage.gui.sidebutton;
+
+import com.raoulvdberge.refinedstorage.gui.GuiBase;
+import com.raoulvdberge.refinedstorage.tile.TileExporter;
+import com.raoulvdberge.refinedstorage.tile.data.TileDataManager;
+import net.minecraft.util.text.TextFormatting;
+
+public class SideButtonRegulator extends SideButton {
+
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/gui/sidebutton/SideButtonRegulator.java]: Unneeded new line

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java [hunk 3]
```diff
@@ -59,7 +78,21 @@ public void updateNode() {
                         ItemStack slot = itemFilters.getStackInSlot(i);
 
                         if (!slot.isEmpty()) {
-                            ItemStack took = network.extractItem(slot, upgrades.getItemInteractCount(), compare, true);
+                            int stackSize = upgrades.getItemInteractCount();
+
+                            if (regulator) {
+                                for(int index = 0; i< handler.getSlots(); i++) {
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: Coding style
- for (
- i < handler.getSlots()

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java [hunk 4]
```diff
@@ -59,7 +78,21 @@ public void updateNode() {
...
+                                    if (slot.isItemEqual(handler.getStackInSlot(index))) {
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: Use API.instance().getComparer().isEqualNoQuantity()

Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: You call `handler.getStackInSlot(index)` you might want to put that in a var

File: src/main/resources/assets/refinedstorage/lang/en_US.lang [hunk 5]
```diff
@@ -143,6 +143,8 @@ sidebutton.refinedstorage:access_type.0=Insert and extract
 sidebutton.refinedstorage:access_type.1=Insert only
 sidebutton.refinedstorage:access_type.2=Extract only
 
+sidebutton.refinedstorage:exporter.regulator=Regulate the export rate
```
Comment (raoulvdberge) [inline src/main/resources/assets/refinedstorage/lang/en_US.lang]: Call this "Regulator Mode"

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java [hunk 6]
```diff
@@ -59,7 +78,21 @@ public void updateNode() {
...
+                                        if (handler.getStackInSlot(index).getCount() >= slot.getCount()) {
+                                            return;
+                                        } else {
+                                            stackSize = upgrades.hasUpgrade(ItemUpgrade.TYPE_STACK) ? slot.getCount() - handler.getStackInSlot(index).getCount() : 1;
```
Comment (way2muchnoise) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: Shouldn't this also break after this?

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

**Comment by raoulvdberge** [comment]
This doesn't seem to handle fluids yet, too.

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java (hunk 6 above)
Comment (InusualZ) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: What if there is multiple stack of the same item/block

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java (hunk 1 above)
Comment (InusualZ) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: If I don't return here would insert the item/block anyway

File: src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java [hunk 7]
```diff
@@ -59,7 +80,22 @@ public void updateNode() {
                         ItemStack slot = itemFilters.getStackInSlot(i);
 
                         if (!slot.isEmpty()) {
-                            ItemStack took = network.extractItem(slot, upgrades.getItemInteractCount(), compare, true);
+                            int stackSize = upgrades.getItemInteractCount();
+
+                            if (regulator) {
+                                for (int index = 0; i < handler.getSlots(); i++) {
+                                    ItemStack handlerStack = handler.getStackInSlot(index);
+                                    if (API.instance().getComparer().isEqualNoQuantity(slot, handlerStack)) {
```
Comment (raoulvdberge) [inline src/main/java/com/raoulvdberge/refinedstorage/tile/TileExporter.java]: Actually, you might wanna use isEqual(slot, handlerStack, compare) here instead.

**Review by way2muchnoise — APPROVED** [review]

**Review by raoulvdberge — APPROVED** [review]

**Pull request merged**